### DISCUSSION TITLE
Get a similarity score in scene matching

### DISCUSSION THREAD

**Discussion by tikitong** [kornia/kornia#1800] [Questions] [answered]
Hello, I'm wondering after reading the tutorial example https://kornia-tutorials.readthedocs.io/en/latest/image_matching.html with image matching. 

Is there a way to get **a unique matching score** to evaluate the correspondence between two images of the same scene? If so, is the method implemented in kornia? 

For example from how many keypoints, with a confidence score at a certain threshold, could we say that the two images represent the same scene? Or again, can we deduce this directly from the number of inliers? Because I imagine that there can be many cases where scenes between two images are different but where numerous inliers points are noted. Thanks for your help.

**Comment by ducha-aiki** [comment, marked as answer]
I can give you theoretical answer and practical answer. 
Theoretical: it is open research question. No number of inliers, no similarity score will guarantee you anything. For example, even simpler question - if this two images are different, or one is an edited copy of the second one, is not yet answered. You can check my recap of the recent competition by Facebook: 
 https://ducha-aiki.github.io/wide-baseline-stereo-blog/2022/02/05/DISC21-recap.html 

Practical: it depends on the local feature and the scene, but basically, 25 inliers after good RANSAC (check this benchmark https://ducha-aiki.github.io/wide-baseline-stereo-blog/2021/05/17/OpenCV-New-RANSACs.html) are good sign that the images are of the same scene.

Before RANSAC? well, I would not trust them. 

So, what I would recommend you, is the following:

1) run LoFTR from kornia, get matches
2) feed them into OpenCV USAC_MAGSAC 
3) check number of inliers > 25

Basically, almost exactly, as implemented in our example  https://kornia-tutorials.readthedocs.io/en/latest/image_matching.html

**Reply by tikitong** [reply]
@ducha-aiki Thank you very much for this answer, and for the two very informative links. After reading through them, I guess your recommendation of 25 inliers after good RANSAC is mostly valid for scenes that I would call "**non-moving**", representing things. For example different shots of a building like in your example, or graphic art on a wall. 

**[** Is your advice still valid for a scene where the background is the same but the foreground changes? For example with two people running in a park and another one representing the same people eating in the same view of the park, we can say that the number of inliers will certainly be >= 25, right? Because of the points of the numerous correspondence between the two backgrounds (trees, park benches, etc). Is RANSAC robust on this point? 

- [x] Sorry question a bit off topic, I just saw that the LoFTR of Kornia was trained on MegaDepth and ScanNet datasets. **]**

But I am still interested in your opinion if RANSAC is robust on this point.

**Reply by ducha-aiki** [reply]
Sorry, I don't understand your question fully. Let me rephrase it:
"Is it possible to use image correspondences to check if the (moving) object is the same in the scene, where the background is the same?"

If this is the question you are asking, then the answer is no. RANSAC in particular and image correspondences in general do not care about background/foreground. 
You probably can do some workarounds, like estimate the background mask and ignore any correspondences from it, etc. 
However, if you are speaking about people, cars, etc, I would recommend something different, like using object detector to detect objects and some metric learning approach to compare bounding boxes. E.g. check this kaggle competition https://www.kaggle.com/c/humpback-whale-identification

**Reply by tikitong** [reply]
Well noted, many thanks @ducha-aiki !
